092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,284 @@
+We will also need to include the fabric8 client dependencies in the notifier JAR that will be used for generating the Kubernetes events as well as for reading the annotation on the Kafka resource from the notifier.
+
+### Pausing the self-healing feature
+
+The user should also have the option to pause the self-healing feature, in case they know or plan to do some rolling restart of the cluster or some rebalance they are going to perform which can cause issue with the self-healing itself.
+For this, we will be having an annotation called `strimzi.io/self-healing-paused`.
+We can set this annotation on the Kafka custom resource to pause the self-healing feature.
+self-healing is not paused by default so missing annotation would mean that anomalies detected will keep getting healed automatically.
+When self-healing is paused, it would mean that all the anomalies that are detected after the pause annotation is applied would be ignored and no fix would be done for them.
+The anomalies that were already getting fixed would continue with the fix.
+
+#### How the pausing works?
+
+To pause self-healing, you will need to annotate the Kafka custom resource with the following annotation:
+
+```sh
+kubectl annotate kafka my-cluster `strimzi.io/self-healing-paused: "true"`
+```
+
+The Kafka custom resource will look like this:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  annotations:
+     strimzi.io/self-healing-paused: "true"
+  creationTimestamp: "2024-12-17T09:01:10Z"
+  generation: 1
+  name: my-cluster
+  namespace: myproject
+  resourceVersion: "109971"
+  uid: 71928731-df71-4b9f-ac77-68961ac25a2f
+```
+
+We will have checks on methods like `onGoalViolation`, `onBrokerFailure`, `onTopicAnomaly` etc. in the `StrimziNotifier` to check if the `pause` annotation is applied on the Kafka custom resource or not.
+If the annotation is applied then this check will set the `action` to be taken on the anomaly as `IGNORE` and the anomalies wouldn't be fixed.
+
+### Strimzi Operations while self-healing is running/fixing an anomaly 
+
+Self-healing is a very robust and resilient feature due to its ability to tackle issues which can appear while the healing of cluster is happening.
+
+#### What happens when self-healing is running/fixing an anomaly and brokers are rolled
+
+If a broker fails, self-healing will trigger and Cruise Control will try to move the partition replicas from the failed broker to other healthy brokers in the Kafka cluster.
+If, during the process of moving partitions from the failed broker, another broker in the Kafka cluster gets deleted or rolled, then Cruise Control would finish the transfer from the failed broker but log that `self-healing finished successful with error`.
```

Comment:
How would it `finish the transfer from the failed broker` when the target broker is stopped? You mean that it will continue with it after the target broker is rolled?

Reply:
Yes, If the target broker is rolling at that moment, then self healing would end saying `self-healing finished successful with error`. Then periodic anomaly check will again detect the `failed broker ` and try to fix it. If the `target` broker is rolled by that time then self healing will start again else if the rolled broker is not up then we will will get two anomalies to fix `the failed broker` and the `target broker`